azkaban-common/src/main/java/azkaban/executor/container/watch/ContainerStatusMetricsHandlerListener.java

```diff
@@ -0,0 +1,190 @@
+  // be 100 bytes * DEFAULT_EVENT_CACHE_MAX_ENTRIES ~ 5 mb
+
+  @Inject
+  public ContainerStatusMetricsHandlerListener(final ContainerMetrics containerMetrics) {
+    this.containerMetrics = containerMetrics;
+
+    this.executor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setNameFormat("azk-container-metrics-pool-%d").build());
+
+    this.podStatusCache = CacheBuilder.newBuilder()
+        .maximumSize(DEFAULT_EVENT_CACHE_MAX_ENTRIES)
+        .recordStats()
+        .build();
+  }
+
+  /**
+   * Checks whether the {@link AzPodStatus} of a new event is the same as the one already cached.
+   *
+   * @param event pod watch event
+   * @return true if the status is different from cached value, false otherwise
+   */
+  protected boolean isUpdatedPodStatusDistinct(final AzPodStatusMetadata event) {
+    AzPodStatus currentStatus = AzPodStatus.AZ_POD_UNSET;
+    currentStatus = podStatusCache.getIfPresent(event.getPodName());
+
+    boolean shouldSkip = currentStatus == event.getAzPodStatus();
+    if (shouldSkip) {
+      logger.info(format("Event pod status is same as current %s, for pod %s."
+          +" Any updates will be skipped.", currentStatus, event.getPodName()));
+    }
+    return !shouldSkip;
+  }
+
+  // Update the cache with the given event.
+  protected void updatePodStatus(final AzPodStatusMetadata event) {
+    podStatusCache.put(event.getPodName(), event.getAzPodStatus());
+    logger.debug(format("Updated status to %s, for pod %s", event.getAzPodStatus(), event.getPodName()));
+  }
+
+  /**
+   * Validate and process event data to emit AzPodStatus metrics
+   * @param event
+   */
+  private void validateAndProcess(final AzPodStatusMetadata event) {
+    requireNonNull(event, "event must be non-null");
+    if (!event.getFlowPodMetadata().isPresent() || !isUpdatedPodStatusDistinct(event)) {
+      return;
+    }
+    // Update AzPodStatus metrics for the flow-pod respectively
+    switch (event.getAzPodStatus()) {
+      case AZ_POD_REQUESTED:
+        containerMetrics.markPodRequested();
+        break;
+      case AZ_POD_SCHEDULED:
+        containerMetrics.markPodScheduled();
+        break;
+      case AZ_POD_INIT_CONTAINERS_RUNNING:
+        containerMetrics.markContainerRunning();
+        break;
+      case AZ_POD_APP_CONTAINERS_STARTING:
+        containerMetrics.markAppContainerStarting();
+        break;
+      case AZ_POD_READY:
+        containerMetrics.markPodReady();
+        break;
+      case AZ_POD_COMPLETED:
+        containerMetrics.markPodCompleted();
+        break;
+      case AZ_POD_INIT_FAILURE:
+        containerMetrics.markPodInitFailure();
+        break;
+      case AZ_POD_APP_FAILURE:
+        containerMetrics.markPodAppFailure();
+        break;
+      default:
+        logger.error(String.format("Current flow pod status %s is not for metric emitting",
+            event.getAzPodStatus()));
```

Comment:
Is this a real error? Or should we just ignore those status for which metrics are not defined?

Reply:
Yes, it's an error, which is unexpected pod status to be logged.